### ISSUE TITLE
Failure to peer despite correct password, when provided by the admin interface.

### ISSUE THREAD

**Issue by cyisfor** [hyperboria/bugs#120]
My netbook and my server simply cannot peer with each other, all of a sudden. Once I went and cleaned up a bit, it appears that the netbook is providing a password that doesn't match what the server expects. But I double checked my "./contrib/python/dynamicEndpoints.py" configuration, and the conf file in the netbook does specify the correct password: an authorized password on my home server.

I checked, and dynamicEndpoints.py is reading the correct password, and the python cjdnsadmin module is bencoding the correct password, and sending it to the socket. So something in the core of cjdns itself is broken, reading the password wrong from the RPC function UDPInterface_beginConnection

Hacking interface/UDPInterface_admin.c to log the password, it is receiving the correct password inside cjdns on my netbook. It is passing the correct password to CryptoAuth_setAuth on my netbook. Then when it tries to peer with my home server... unrecognized auth.

**Comment by cyisfor** [comment]
So, github sucks and "Something went really wrong, and we can’t process that file." I'll just copy and paste it here.

```
On server:
1425945507 DEBUG SessionManager.c:379 Session sendPath[0013.cb8b.8ace.8a07] recvPath[0013.cb8b.8ace.8a07] ip[fc3b:dc68:2f4d:7c7d:abbc:c32c:3541:6b8d] ended

1425946189 DEBUG SessionManager.c:276 ver[0] send[0] recv[65377] ip[fc3b:dc68:2f4d:7c7d:abbc:c32c:3541:6b8d] path[0027.9717.07c8.74e5] new session nonce[1]
1425946189 DEBUG CryptoAuth.c:591 0x55cbb872bc18 inner [fc3b:dc68:2f4d:7c7d:abbc:c32c:3541:6b8d]: Received a repeat hello packet
1425946189 DEBUG CryptoAuth.c:642 0x55cbb872bc18 inner [fc3b:dc68:2f4d:7c7d:abbc:c32c:3541:6b8d]: DROP message with nonce [1], decryption failed
1425946189 DEBUG SessionManager.c:284 ver[0] send[0] recv[65377] ip[fc3b:dc68:2f4d:7c7d:abbc:c32c:3541:6b8d] path[0027.9717.07c8.74e5] DROP Failed decrypting message NoH[1] state[CryptoAuth_NEW]

On laptop:

1425946288 DEBUG InterfaceController.c:350 Pinging unresponsive peer [nhmsdzjv6p2gmmwgl8tjlxkj2y0l1b382ur67ppgvnw3pxs96zf0.k] lag [412]
1425946288 DEBUG InterfaceController.c:286 SwitchPing [nhmsdzjv6p2gmmwgl8tjlxkj2y0l1b382ur67ppgvnw3pxs96zf0.k]
1425946288 DEBUG CryptoAuth.c:393 0x555c1bc22da8 outer [fc38:7233:d541:c707:d8d5:8a60:4da9:78a9]: Sending repeat hello packet
1425946288 DEBUG Pathfinder.c:312 Peer gone [v0.0000.0000.0000.0013.nhmsdzjv6p2g
mmwgl8tjlxkj2y0l1b382ur67ppgvnw3pxs96zf0.k]
1425946289 DEBUG Pinger.c:73 Ping timeout for [1528990115] in [2048] ms
1425946289 DEBUG InterfaceController.c:350 Pinging lazy peer [lhjs0njqtvh1z4p2922bbyp2mksmyzf5lb63kvs3ppy78y1dj130.k] lag [18]
1425946289 DEBUG InterfaceController.c:286 SwitchPing [lhjs0njqtvh1z4p2922bbyp2mksmyzf5lb63kvs3ppy78y1dj130.k]
1425946289 DEBUG CryptoAuth.c:393 0x555c1bc43458 outer [fcd9:e703:498e:5d07:e5fc
:d525:80a6:a51c]: Sending repeat hello packet
1425946290 WARN Janitor.c:609 No nodes in routing table, check network connectio
n and configuration.


1425946358 DEBUG InterfaceController.c:350 Pinging unresponsive peer [nhmsdzjv6p
2gmmwgl8tjlxkj2y0l1b382ur67ppgvnw3pxs96zf0.k] lag [482]
1425946358 DEBUG InterfaceController.c:286 SwitchPing [nhmsdzjv6p2gmmwgl8tjlxkj2
y0l1b382ur67ppgvnw3pxs96zf0.k]
1425946358 DEBUG CryptoAuth.c:393 0x555c1bc22da8 outer [fc38:7233:d541:c707:d8d5
:8a60:4da9:78a9]: Sending repeat hello packet
1425946358 DEBUG Pathfinder.c:312 Peer gone [v0.0000.0000.0000.0013.nhmsdzjv6p2g
mmwgl8tjlxkj2y0l1b382ur67ppgvnw3pxs96zf0.k]
1425946359 DEBUG InterfaceController.c:736 sendBeacon(UDP) -> beaconing disabled

1425946359 DEBUG Pinger.c:73 Ping timeout for [1528990176] in [2048] ms
1425946359 DEBUG InterfaceController.c:350 Pinging unresponsive peer [lhjs0njqtvh1z4p2922bbyp2mksmyzf5lb63kvs3ppy78y1dj130.k] lag [20]
1425946359 DEBUG InterfaceController.c:286 SwitchPing [lhjs0njqtvh1z4p2922bbyp2mksmyzf5lb63kvs3ppy78y1dj130.k]
1425946359 DEBUG CryptoAuth.c:393 0x555c1bc3e718 outer [fcd9:e703:498e:5d07:e5fc
:d525:80a6:a51c]: Sending repeat hello packet
1425946359 DEBUG Pathfinder.c:312 Peer gone [v0.0000.0000.0000.00d2.lhjs0njqtvh1
z4p2922bbyp2mksmyzf5lb63kvs3ppy78y1dj130.k]
1425946359 DEBUG UDPInterface_admin.c:53 Peering with [lhjs0njqtvh1z4p2922bbyp2m
ksmyzf5lb63kvs3ppy78y1dj130.k]
1425946359 DEBUG InterfaceController.c:826 bootstrapPeer total [2]
1425946359 INFO InterfaceController.c:877 Adding peer [v0.0000.0000.0000.00d6.lh
js0njqtvh1z4p2922bbyp2mksmyzf5lb63kvs3ppy78y1dj130.k] from bootstrapPeer()
1425946359 DEBUG InterfaceController.c:286 SwitchPing [lhjs0njqtvh1z4p2922bbyp2m
ksmyzf5lb63kvs3ppy78y1dj130.k]
1425946359 DEBUG CryptoAuth.c:393 0x555c1bc4b1a8 outer [fcd9:e703:498e:5d07:e5fc
:d525:80a6:a51c]: Sending hello packet
```

**Comment by cyisfor** [comment]
I'm going to close this, because I went and manually wrote a custom cjdroute.conf file with the current IP address written into it, and a valid password, and it still failed to authenticate. Dunno why this is happening all of a sudden. They're both the exact same version, and I've never had any authentication problem before.

**Comment by cyisfor** [comment]
Ugh... okay, I made two configurations the first one with the password that connects to my home server, and the second with an empty connectTo.

* Running the first one, it connects fine, and peerStats reports ESTABLISHED on both sides. 
* Killing that, and running the second one, the netbook has no peers initially (obviously) and the home server drops to UNRESPONSIVE. Then when I run dynamicEndpoints.py with the _exact same password_ the netbook adds my home server as a peer, which goes unresponsive, while my home server reports authentication failures.
* Waiting for my home server to report zero peers, then running the second one, same deal.
* Killing that and running the first, they both immediately become ESTABLISHED again.

So, exact same password. Using the admin interface (on the netbook side) causes an authentication error.

**Comment by cyisfor** [comment]
Okay, I figured out the problem, at least. Is anyone even working on cjdns anymore? Victim of its own code stability? 

Well, cjdns tells the "cjdnsadmin.py" module "login is an optional string argument" and the module interprets that as "specify a string consisting of the character "0" as a login argument every time."

And I have no clue what "login" is, but it overrides any checking based on a shared secret password. So, no matter what password you specify, since login is wrongly provided, authType will get set to 2, and it will only check the login field, which cjdnsadmin.py also wrongly sets to "0".

**Comment by lgierth** [comment]
@cyisfor great that you found the issue -- people do still work on cjdns and we're happy to merge your patch fixing the issue! I'm not aware of anyone actually using the python stuff, so we rarely have the capacity to fix it (or even notice issues).

**Comment by cyisfor** [comment]
No problem. I think it's just cjdns is such a stable code base that there's so little work to be done on it, compared to sexier bug-ridden programs that you can go be a hero on.

How do you use the cjdns admin interface? I'll fix the python, but being able to set up a connection to a dynamic IP is what I would like to do, and if everyone's using another language I'd be happy to do it there instead.

**Comment by lgierth** [comment]
The stuff in `tools/` is much nicer than anything in `contrib/` then, and best is it's maintained :) Check out `tools/cexec` in particular, which lets you make direct calls to the API, e.g. `tools/cexec "InterfaceController_peerStats()"`

**Comment by cyisfor** [comment]
Thanks! Also, success!

**Comment by cyisfor** [comment]
Once [this](https://github.com/hyperboria/cjdns/pull/121) pull request gets merged, can close this issue.

**Comment by MaStr** [comment]
was merged into 05a29812e7f659d9e6495f65314dac463f1d98f3 on upstream branch crashey 
can be closed
